src/Storages/ObjectStorage/DataLakes/DeltaLake/KernelHelper.cpp

```diff
@@ -26,29 +26,51 @@ class S3KernelHelper final : public IKernelHelper
 public:
     S3KernelHelper(
         const DB::S3::URI & url_,
-        const std::string & access_key_id_,
-        const std::string & secret_access_key_,
-        const std::string & region_,
-        const std::string & token_,
-        bool no_sign_)
+        std::shared_ptr<const DB::S3::Client> client_,
+        const DB::S3::S3AuthSettings & auth_settings)
         : url(url_)
-        , access_key_id(access_key_id_)
-        , secret_access_key(secret_access_key_)
-        , region(region_)
-        , token(token_)
-        , no_sign(no_sign_)
         , table_location(getTableLocation(url_))
+        , client(client_)
     {
         /// Check if user didn't mention any region.
         /// Same as in S3/Client.cpp (stripping len("https://s3.")).
         if (url.endpoint.substr(11) == "amazonaws.com")
             url.addRegionToURI(region);
+
+        const auto & credentials = client->getCredentials();
+        access_key_id = credentials.GetAWSAccessKeyId();
+        secret_access_key = credentials.GetAWSSecretKey();
+        token = credentials.GetSessionToken();
+
+        region = client->getRegion();
+        if (region.empty() || region == Aws::Region::AWS_GLOBAL)
+            region = client->getRegionForBucket(url.bucket, /* force_detect */true);
+
+        no_sign = auth_settings[DB::S3AuthSetting::no_sign_request];
     }
 
     const std::string & getTableLocation() const override { return table_location; }
 
     const std::string & getDataPath() const override { return url.key; }
 
+    bool update() override
```

Comment:
Why not just take credentials from the client directly whenever `createBuilder` is called?